CCEffectRefraction - Fix various issues uncovered by Viktor during testing

thayerandrews opened the pull request:
- Use premultiplied alpha blending to combine the sprite's main texture with the refracted environment.
- If the refracted texture coordinates would be out of bounds of the environment, do not combine the resulting environment pixel with the sprite's main texture.
- Fix a crash that occurred if an effect was applied to a sprite with a nil main texture.
- If a sprite has no main texture, set the texture rect from the normal map instead.

Commit:
CCEffectRefraction - Address feedback from Viktor

- Fix the blend of the main texture and the refracted environment texture. It was a multiply, now it's a
  premultiplied alpha blend.
- If the refracted texture coordinates would index outside the environment texture, just return the main
  texture.

Commit:
CCEffectRefraction - Crash fix when sprite texture / sprite frame is nil

Commit:
CCSprite - Set the sprite's texture rect from the normal map if there is no main texture.

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

slembcke commented on cocos2d/CCEffectRefraction.m:
```diff
@@ -74,7 +74,12 @@ -(void)buildFragmentFunctions
 
                                    vec4 primaryColor = cc_FragColor * texture2D(cc_MainTexture, cc_FragTexCoord1);
-                                   return primaryColor * texture2D(u_envMap, refractTexCoords);
+                                   if ((refractTexCoords.x >= 0.0) && (refractTexCoords.x <= 1.0) &&
```
I'm always a little leery of using conditionals in fragment shaders. On pre-A7 devices I've done something like this:

```
vec2 compare = 0.5 - abs(refractTexCoords - 0.5)
primaryColor += step(min(compare.x, compare.y), 0.0)*texture2D(...)
```

thayerandrews replied:
Nice avoidance of the if statement. And I agree I'd rather not introduce the conditional if it can be avoided which apparently it can quite easily. :)

Commit:
CCEffectRefraction - Remove conditional based on the refracted texture coordinates

Instead use abs(), min(), and step() to arrive at a flag that is 1.0 if the refracted coords are in bounds and
0.0 if they are out. Thanks to Scott for pointing out this alternative solution.

thayerandrews commented:
Updated to include Scott's optimization above. The only difference is that GLSL's step() function is step(edge, value_to_compare_to_edge) so the args to that function are swapped when compared to Scott's example.

Commit:
Merge remote-tracking branch 'upstream/develop' into develop
